### PR TITLE
Fix alpha-substitution for nested lambdas in a generic context.

### LINKED ISSUE #2549: [BUG] Roslyn outputs incorrect IL according to PEVERIFY for this test case
The following class when compiled with VS2015rc generates bad IL according to PEVERIFY

	    public class BadBaby
	    {
	        IEnumerable<bool> Children;
	
	        public IEnumerable<object> Foo<TArtifact>()
	        {
	
	            return Enumerable
	                    .Empty<bool>()
	                    .Select(_ => Children
	                        .OfType<TArtifact>()
	                        .Select(c => false)
	                    )
	                ;
	        }
	
	    }
	    
The above minimal code was extracted through trial and error from a much more complex part of our production code.
	    
output from PEVERIFY

	C:\Users\phelan\workspace\BadBadBaby\BadBadBaby\bin\Debug
	> PEVerify.exe .\BadBadBaby.dll
	
	Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.33440
	Copyright (c) Microsoft Corporation.  All rights reserved.
	
	[IL]: Error: [C:\Users\phelan\workspace\BadBadBaby\BadBadBaby\bin\Debug\BadBadBaby.dll : 
	    BadBadBaby.BadBaby::<Foo>b__1_0[TArtifact]][offset 0x0000000B] 
	     [HRESULT 0x8007000B] - 
	 An attempt was made to load a program with an incorrect format.
	
	1 Error(s) Verifying .\BadBadBaby.dll
	C:\Users\phelan\workspace\BadBadBaby\BadBadBaby\bin\Debug
	>
	
The complete project complete with built assemblies can be found at

https://www.dropbox.com/s/mw6x2r4uui54soy/BadBadBaby.zip?dl=0

### PR DESCRIPTION
**Pull request by gafter**
Fixes #2549 
@agocke @AlekseyTs @VSadov please review

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs [hunk 1]
```diff
@@ -1171,7 +1171,7 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
                     BoundExpression cache;
                     if (shouldCacheForStaticMethod || shouldCacheInLoop && (object)containerAsFrame != null)
                     {
-                        var cacheVariableType = containerAsFrame?.TypeMap.SubstituteType(type) ?? translatedLambdaContainer;
+                        var cacheVariableType = containerAsFrame?.TypeMap.SubstituteType(node.Type) ?? translatedLambdaContainer;
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: We are using un-rewritten node.Type here. Other places use rewritten type. Seems very subtle. Could we have a small comment?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs]: Sure.

**Comment by VSadov** [comment]
@gafter - thanks! 
LGTM

**Comment by gafter** [comment]
@VSadov The line you asked me to comment on was changed by you from

```cs
                        var cacheVariableType = lambdaIsStatic ? type : (translatedLambdaContainer as LambdaFrame).TypeMap.SubstituteType(type);
```
to
```cs
                        var cacheVariableType = containerAsFrame?
                            .TypeMap.SubstituteType(type) ?? 
                            translatedLambdaContainer;
```
in a change on 10/22/2014 whose comment was
```none
Making lambda codegen to always emit instance delegates.

This way we can have better performance of invocations while keeping parameter list of delegate.Method backwards compatible.

The approach is basically always to emit backing methods into a display class, just that in a case of lambdas not capturing anything, the display classes will not have any fields.
Also, when display classes do not have any fields, lambdas of all nongeneric methods in a given class can share the same display class and in fact can share the same singleton instance.

***NO_CI***
 (changeset 1359290)
```
I cannot understand this change. I don’t see how “translaterLambdaContainer” can ever be the correct type for this variable. Should this line just be
```cs
                        var cacheVariableType = containerAsFrame.TypeMap.SubstituteType(node.Type);
```
?

**Comment by VSadov** [comment]
@gafter  - I think I was somehow convinced that we have to use rewritten "type" everywhere (which as you explain in this fix is not correct) and that led me on the path of using translaterLambdaContainer when container is the actual containing type.

Yes, what you suggest seems more appropriate and also simpler.

**Comment by VSadov** [comment]
@gafter - could you check if we have same issue on the VB side? Perhaps need an issue to track that.

**Comment by gafter** [comment]
VB does not have this precise issue.

**Pull request merged**